source/distributed/tria_base.cc

```diff
@@ -342,6 +346,201 @@ namespace parallel
+      else
+        cell->set_global_active_cell_index(numbers::invalid_dof_index);
+
+    // 4) determine the global indices of ghost cells
+    GridTools::exchange_cell_data_to_ghosts<types::global_cell_index>(
+      *this,
+      [](const auto &cell) { return cell->active_cell_index(); },
+      [](const auto &cell, const auto &id) {
+        cell->set_global_active_cell_index(id);
+      });
+
+    // 5) set up new partitioner
+    IndexSet is_local(this->n_global_active_cells());
+    IndexSet is_ghost(this->n_global_active_cells());
+
+    for (const auto &cell : this->active_cell_iterators())
+      if (!cell->is_artificial())
+        {
+          const auto index = cell->active_cell_index();
+
+          if (index == numbers::invalid_dof_index)
+            continue;
+
+          if (cell->is_locally_owned())
+            is_local.add_index(index);
+          else
+            is_ghost.add_index(index);
+        }
+
+    number_cache.active_cell_index_partitioner =
+      Utilities::MPI::Partitioner(is_local, is_ghost, this->mpi_communicator);
+
+    // 6) translate global to local indices
+    for (const auto &cell : this->active_cell_iterators())
+      if (!cell->is_artificial())
+        {
+          if (cell->active_cell_index() != numbers::invalid_dof_index)
+            cell->set_global_active_cell_index(
+              number_cache.active_cell_index_partitioner.global_to_local(
+                cell->active_cell_index()));
+        }
+
+    // 7) proceed with multigrid levels if requested
+    if (this->is_multilevel_hierarchy_constructed() == true)
+      {
+        // 1) determine number of locally-owned cells on levels
+        std::vector<types::global_cell_index> n_locally_owned_cells(
+          this->n_global_levels(), 0);
+
+        for (auto cell : this->cell_iterators())
+          if (cell->level_subdomain_id() == this->locally_owned_subdomain())
+            n_locally_owned_cells[cell->level()]++;
+
+        // 2) determine the offset of each process
+        std::vector<types::global_cell_index> cell_index(
+          this->n_global_levels(), 0);
+
+        MPI_Exscan(n_locally_owned_cells.data(),
+                   cell_index.data(),
+                   this->n_global_levels(),
+                   Utilities::MPI::internal::mpi_type_id(
+                     n_locally_owned_cells.data()),
+                   MPI_SUM,
+                   this->mpi_communicator);
+
+        // 3) determine global number of "active" cells on each level
+        std::vector<types::global_cell_index> n_cells_level(
+          this->n_global_levels(), 0);
+
+        MPI_Allreduce(n_locally_owned_cells.data(),
+                      n_cells_level.data(),
+                      this->n_global_levels(),
+                      Utilities::MPI::internal::mpi_type_id(
+                        n_locally_owned_cells.data()),
+                      MPI_SUM,
+                      this->mpi_communicator);
```

Comment:
I'm sure the additional efficiency is minuscule compare to , but I ask nonetheless: What would you think of doing an `MPI_Bcast` the result from `ExScan` on the last rank, augmented by the sum of cells on all levels on that rank?